### PR TITLE
PHOENIX-2827 Support OFFSET in Calcite-Phoenix

### PR DESCRIPTION
**Pull request by lomoree**

### CODE REVIEW

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java [hunk 1]
```diff
@@ -76,20 +76,27 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
     
     @Override 
     public double estimateRowCount(RelMetadataQuery mq) {
-        double rows = super.estimateRowCount(mq);        
+        double rows = super.estimateRowCount(mq);
+        if(offset != null) {
+            return Math.max(0, Math.min(RexLiteral.intValue(fetch), rows - RexLiteral.intValue(offset)));
+        }
         return Math.min(RexLiteral.intValue(fetch), rows);
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java]: I'd write:
int offset = this.offset == null ? 0 : RexLiteral.intValue(this.offset);
return Math.min(RexLiteral.intValue(fetch), rows - offset);

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java [hunk 2]
```diff
@@ -76,20 +76,27 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
...
 
     @Override
     public QueryPlan implement(PhoenixRelImplementor implementor) {
         QueryPlan plan = implementor.visitInput(0, (PhoenixQueryRel) getInput());
         int fetchValue = RexLiteral.intValue(fetch);
+        int offsetValue = 0;
+        if (offset != null){
+            offsetValue = RexLiteral.intValue(offset);
+        }
         if (plan.getLimit() == null) {
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java]: should be "if (plan.getLimit() == null && plan.getOffset() == null)"

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java [hunk 3]
```diff
@@ -76,20 +76,27 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
...
+        if(offset != null) {
+            return Math.max(0, Math.min(RexLiteral.intValue(fetch), rows - RexLiteral.intValue(offset)));
+        }
         return Math.min(RexLiteral.intValue(fetch), rows);
     }
 
     @Override
     public QueryPlan implement(PhoenixRelImplementor implementor) {
         QueryPlan plan = implementor.visitInput(0, (PhoenixQueryRel) getInput());
         int fetchValue = RexLiteral.intValue(fetch);
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java]: I believe now that we've enabled offset, "this.fetch" could be also be null, which means it should be handled the same way as this.offset now. Could you please also add test cases to cover situations like "offset != null" but "limit == null"?

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java [hunk 4]
```diff
@@ -76,20 +76,27 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
...
+        }
         if (plan.getLimit() == null) {
-            return plan.limit(fetchValue);
+            return plan.limit(fetchValue, offsetValue);
         }
         
         return new ClientScanPlan(plan.getContext(), plan.getStatement(), 
                 implementor.getTableMapping().getTableRef(), RowProjector.EMPTY_PROJECTOR, 
-                fetchValue, null, null, OrderBy.EMPTY_ORDER_BY, plan);
+                fetchValue, offsetValue, null, OrderBy.EMPTY_ORDER_BY, plan);
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rel/PhoenixLimit.java]: Since either of them could be null, Boolean objects should be used instead of primitive boolean type.

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rules/PhoenixConverterRules.java [hunk 5]
```diff
@@ -232,8 +232,7 @@ public RelNode convert(RelNode rel) {
         private static Predicate<LogicalSort> HAS_FETCH = new Predicate<LogicalSort>() {
             @Override
             public boolean apply(LogicalSort input) {
-                return input.offset == null 
-                        && input.fetch != null;
+                return input.fetch != null;
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rules/PhoenixConverterRules.java]: should be "input.fetch != null || input.offset != null"

File: phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java [hunk 6]
```diff
@@ -106,12 +106,18 @@ public ExplainPlan getExplainPlan() throws SQLException {
     }
 
     @Override
-    public QueryPlan limit(Integer limit) {
-        if (limit == this.limit || (limit != null && limit.equals(this.limit)))
+    public QueryPlan limit(Integer limit, Integer offset) {
+        if (limit == this.limit || (limit != null && limit.equals(this.limit))) {
```
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java]: I can see there's a big confusion here. Basically, offset and limit should be treated the same way. So here, it should be "if both the old limit and the old offset are same as as the new ones specified, we return this; otherwise we create a new object."

Comment (lomoree) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java]: This does definitely have ripple effects throughout, but I had done this on the basis that SQL only allows an offset if their is an imposed limit (since an offset doesn't make any logical sense without a limit).

File: phoenix-core/src/main/java/org/apache/phoenix/calcite/rules/PhoenixConverterRules.java (hunk 5 above)
Comment (lomoree) [inline phoenix-core/src/main/java/org/apache/phoenix/calcite/rules/PhoenixConverterRules.java]: With respect to my comment below, if there is a limit, then we apply the rule, otherwise we do not (regardless of offset).
 I may have the wrong idea here, but based on my understanding of offset, all of the test cases pass as expected. Are the test cases I have laid out in CalciteIT what you would expect?

File: phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java (hunk 6 above)
Comment (maryannxue) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java]: Is it a SQL-92 standard that OFFSET can only be used together with LIMIT? I know there are some dialects that allow different OFFSET/LIMIT or OFFSET/FETCH grammars, so I'm just confused... I tried with Calcite parser and it does allow OFFSET without LIMIT.

Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java]: For right or wrong, Phoenix allows an OFFSET to be used without a LIMIT today.

Comment (lomoree) [inline phoenix-core/src/main/java/org/apache/phoenix/execute/ClientScanPlan.java]: Ah, my mistake. I'll get working on a patch that also supports offset without limit. This should incorporate all of the changes noted above. Thanks for clarifying.
